### PR TITLE
techdocs: Expose backend 500 errors (mkdocs build) to UI

### PR DESCRIPTION
**Pull request by adamdmharvey**
## Hey, I just made a Pull Request!

An attempt to address #4111.

* Updates the `<TechDocsNotFound />` component to support passing an error code. Was hardcoded to 404, but allow it to be passed in now. It's Tech Docs Not Found, but maybe not FILE not found, per se.
* Updates API to trap 500 errors, and change the error messaging.
Now, if an mkdocs build error has happened which is different than a repo lacks an index.md, the user will see this error:

image

Note: I was unable to figure out how to pass the HTTP error code from the API back into the component, so while the React UI component COULD say ERROR 500, it doesn't yet. Happy for feedback on that, or to iterate later.

This helps make progress towards my linked issue #4111, and potentially closes it, but open to opinions on that.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/techdocs/src/api.ts
Comment (adamdmharvey) [inline plugins/techdocs/src/api.ts:133]: Welcome any proposals for other wording here? I suppose it fits my use case, but a 500 error could be anything. I'm referring to a generator message, but if someone is using an external generator, this could also become a preparer or publisher error, so something even _more generic_ but that does not infer "file not found" could be used.

Comment (OrkoHunter) [reply plugins/techdocs/src/api.ts:133]: We can do "Could not generate documentation or something else went wrong in (backend/server)."

In `TechDocsNotFound` component, there is an additional message based on `techdocsBuilder !== 'local'`, which helps.

**Review by OrkoHunter — APPROVED** [review]
Thanks! 🙏

File: plugins/techdocs/src/reader/components/TechDocsNotFound.tsx
Comment (OrkoHunter) [inline plugins/techdocs/src/reader/components/TechDocsNotFound.tsx:41]: > Note: I was unable to figure out how to pass the HTTP error code from the API back into the component, so while the React UI component COULD say ERROR 500, it doesn't yet. Happy for feedback on that, or to iterate later.

I thought the ErrorPage from `@backstage/core` is responsible for the error code. Sorry, I don't think I understand "it doesn't yet" part!

Comment (adamdmharvey) [reply plugins/techdocs/src/reader/components/TechDocsNotFound.tsx:41]: Yeah sorry, this component `<TechDocsNotFound>` receives the status, and so it gets passed into the error code.  The challenge I had was elsewhere... I'll separately comment where exactly I meant.

File: plugins/techdocs/src/reader/components/Reader.tsx
Comment (adamdmharvey) [inline plugins/techdocs/src/reader/components/Reader.tsx:158]: At this point, a shadowDom has been calculcated after being received back from the API call.  I could not figure out how to get the statusCode from the resulting API call, passed through to this method here, and therefore to change this line.  Ideally, it should have been something like:

```typescript
return <TechDocsNotFound statusCode={returnedApiStatusCode} errorMessage{errror.message} />;
```

But since I couldn't collect it, I just hardcoded it as the 404 into this component.  If you had any suggestions for this Reader method I'd welcome it!

Comment (adamdmharvey) [reply plugins/techdocs/src/reader/components/Reader.tsx:158]: This may be where my lack of TypeScript expertise as a non-programmer rears it's ugly face. I guess some way to collect from this call?

https://github.com/backstage/backstage/blob/45de779d5fa75721dd8bb1865c3e63a8ba8c4976/plugins/techdocs/src/reader/components/Reader.tsx#L53-L55

Comment (OrkoHunter) [reply plugins/techdocs/src/reader/components/Reader.tsx:158]: Yeah! IMO we should be throwing separate custom error types from the API logic
https://github.com/backstage/backstage/blob/45de779d5fa75721dd8bb1865c3e63a8ba8c4976/plugins/techdocs/src/api.ts#L124-L135

and use the error type to display the right error message. I am not sure if the `@backstage` frontend package provides custom error types like the backend has 

https://github.com/backstage/backstage/blob/435972704c9acf85fda79e28c9556482780c801d/packages/backend-common/src/errors.ts#L50-L82

@Rugvip @freben I'll be thankful for your comments on this!

Comment (adamdmharvey) [reply plugins/techdocs/src/reader/components/Reader.tsx:158]: Wonder if this might be ok for an incremental improvement without the custom error throws, and then increment later. Key for me is avoiding just the raw msg being printed like so:

image

And this PR does accomplish that?  I could add a comment into where I hardcode the 404 and say why with a TODO? What do you think @OrkoHunter ?

Comment (OrkoHunter) [reply plugins/techdocs/src/reader/components/Reader.tsx:158]: A TODO about 404 sounds fine @adamdmharvey. :)

**Comment by adamdmharvey** [comment]
Have a feeling this is about to conflict with #4401.

**Comment by OrkoHunter** [comment]
Let's get this PR in and rebase the other one. cc @emmaindal 

We can skip the maintainers' review if you prefix your changeset with `techdocs-`. Or if @backstage/maintainers have time, this one needs a review! :)

**Comment by adamdmharvey** [comment]
Thanks, @OrkoHunter ! Totally forgot about `.changeset/techdocs-*`. Done and added TODO.

**Comment by emmaindal** [comment]
> Let's get this PR in and rebase the other one. cc @emmaindal
> 

yesss! 🚢

**Comment by adamdmharvey** [comment]
When this goes green, if it's only FOSSA that failed, and given Orko's approval, I'll merge this in if that's ok and then Emma you can merge w/ your branch.

**Comment by OrkoHunter** [comment]
:shipit:

**Comment by adamdmharvey** [comment]
E2E Linux test at 47 mins now.... it averages 15. Hmm. 🕙

**Comment by adamdmharvey** [comment]
Hmm.... @OrkoHunter if you're admin, are you able to re-trigger the workflow for E2E Linux when you get a chance? if not i can just push one more commit to force it also.

> App serve failed, trying again, Error: Timeout: did not get to load all resources on this page

**Comment by emmaindal** [comment]
retriggered!

**Comment by OrkoHunter** [comment]
Thanks Emma!

**Comment by OrkoHunter** [comment]
Wow, even FOSSA passed. Please merge before it changes its mind.

**Comment by adamdmharvey** [comment]
thanks all 🤗

**Pull request merged**
